File: mycu/core/session.py
"""Session state: where it lives on disk, and what little of it we own.

An important thing to be clear about: **this module does not store cookies.**
The session cookie is `HttpOnly`, and on Android it is unreachable from Qt (the
`cookieAdded` signal only fires for cookies *we* set, which is exactly why the
transport lets the WebView make the request instead of harvesting the cookie).
The cookie jar is owned by the WebView:

* **Desktop** — QtWebEngine's persistent profile, stored under
  :meth:`SessionStore.profile_dir`.
* **Android** — the system WebView's own cookie store, in app-private storage.
  We neither see it nor manage it.

What this module *does* own is the small amount of metadata that makes the UI
behave sensibly across launches: have we ever logged in, when did a request last
succeed, and which term were we last looking at. It is written 0600 under
``$XDG_DATA_HOME`` (or ``filesDir`` on Android) and never goes near git.

The app never sees or stores your password. Login happens on Microsoft's own
page, inside the WebView.
"""

from __future__ import annotations

import json
import os
import time
from dataclasses import asdict, dataclass
from pathlib import Path

APP_DIR_NAME = "mycu"

#: Bumped whenever the on-disk shape of ``session.json`` changes. A file with a
#: different value is discarded rather than migrated — it holds nothing
#: expensive to regenerate.
SCHEMA_VERSION = 1


@dataclass(slots=True)
class SessionState:
    """Non-secret metadata about the current session.

    Nothing here is a credential. If this file leaked it would reveal that you
    use the app and roughly when — which is why it is still written 0600, but
    also why losing it costs you nothing more than one extra login.
    """

    #: Unix time of the last completed interactive login, 0 if never.
    last_login: float = 0.0
    #: Unix time of the last request that came back with real data.
    last_success: float = 0.0
    #: Unix time we last detected expiry, for "your session ended" messaging.
    last_expiry: float = 0.0
    #: Whatever the chapel view was last showing, so a cold start looks right.
    last_term: str = ""
    #: Copy of :data:`SCHEMA_VERSION` at write time; see :meth:`SessionStore.load`.
    schema: int = SCHEMA_VERSION

    @property
    def has_logged_in(self) -> bool:
        """Whether we have ever completed a login on this device.

        Drives first-run behaviour only. It is *not* a claim that the session is
        still valid — that is never predicted, only discovered by making a
        request. See :func:`mycu.core.transport.looks_like_login`.
        """
        return self.last_login > 0


def default_state_dir() -> Path:
    """Where session state belongs on this platform.

    Android is detected via ``ANDROID_ARGUMENT``, which python-for-android sets
    in every app process; ``MYCU_STATE_DIR`` overrides both for tests and for
    running two profiles side by side.
    """
    override = os.environ.get("MYCU_STATE_DIR")
    if override:
        return Path(override)

    android_private = os.environ.get("ANDROID_PRIVATE") or os.environ.get("ANDROID_ARGUMENT")
    if android_private:
        # ANDROID_ARGUMENT points at the app's private data dir; ANDROID_PRIVATE
        # is the same thing on newer python-for-android. Either is app-private
        # and unreadable by other apps, which is what we want.
        return Path(android_private) / APP_DIR_NAME

    xdg = os.environ.get("XDG_DATA_HOME")
    base = Path(xdg) if xdg else Path.home() / ".local" / "share"
    return base / APP_DIR_NAME


class SessionStore:
    """Loads, saves and clears :class:`SessionState`.

    Deliberately forgiving on read: a corrupt or older-schema file is treated as
    "no session", because the cost of being wrong is one login, and the cost of
    crashing on startup is an app that cannot recover without a terminal.
    """

    FILENAME = "session.json"

    def __init__(self, state_dir: Path | str | None = None) -> None:
        self.state_dir = Path(state_dir) if state_dir else default_state_dir()

    @property
    def path(self) -> Path:
        return self.state_dir / self.FILENAME

    def profile_dir(self) -> Path:
        """Directory for the WebView's persistent cookie jar (desktop only).

        Handed to ``QWebEngineProfile.setPersistentStoragePath``. Android
        ignores this entirely — the system WebView manages its own store.
        """
        return self.state_dir / "webprofile"

    def ensure_dirs(self) -> None:
        """Create the state directory 0700 if it is not already there."""
        self.state_dir.mkdir(parents=True, exist_ok=True, mode=0o700)

    def load(self) -> SessionState:
        try:
            raw = json.loads(self.path.read_text(encoding="utf-8"))
        except (FileNotFoundError, json.JSONDecodeError, OSError):
            return SessionState()

        if not isinstance(raw, dict) or raw.get("schema") != SCHEMA_VERSION:
            return SessionState()

        known = set(SessionState.__slots__)
        return SessionState(**{k: v for k, v in raw.items() if k in known})

    def save(self, state: SessionState) -> None:
        """Write state atomically and 0600.

        Atomic because a half-written file on a crash would otherwise be read
        back as "no session" — recoverable, but an avoidable surprise.
        """
        self.ensure_dirs()
        tmp = self.path.with_suffix(".json.tmp")
        tmp.write_text(json.dumps(asdict(state), indent=2), encoding="utf-8")
        os.chmod(tmp, 0o600)
        tmp.replace(self.path)

    def mark_login(self, state: SessionState) -> SessionState:
        state.last_login = time.time()
        self.save(state)
        return state

    def mark_success(self, state: SessionState) -> SessionState:
        state.last_success = time.time()
        self.save(state)
        return state

    def mark_expiry(self, state: SessionState) -> SessionState:
        state.last_expiry = time.time()
        self.save(state)
        return state

    def clear(self) -> None:
        """Forget the session metadata.

        Does **not** clear cookies — that is the WebView's job, and the UI does
        it through :meth:`mycu.ui.login.LoginController.clear_cookies` so both
        halves are dropped together.
        """
        self.path.unlink(missing_ok=True)

File: src/core/session.cpp
#include "session.h"

#include "log.h"

#include <QDateTime>
#include <QDir>
#include <QFile>
#include <QJsonDocument>
#include <QJsonObject>
#include <QStandardPaths>

#include <cstdio>

namespace mycu {

namespace {

double now()
{
    return QDateTime::currentMSecsSinceEpoch() / 1000.0;
}

} // namespace

QString defaultStateDir()
{
    const QString override = qEnvironmentVariable("MYCU_STATE_DIR");
    if (!override.isEmpty())
        return override;

#ifdef Q_OS_ANDROID
    // `<APPROOT>/files` on Android: app-private and unreadable by other apps,
    // which is what we want.
    return QDir(QStandardPaths::writableLocation(QStandardPaths::AppDataLocation))
        .filePath(APP_DIR_NAME);
#else
    const QString xdg = qEnvironmentVariable("XDG_DATA_HOME");
    const QString base = !xdg.isEmpty() ? xdg : QDir::home().filePath(QStringLiteral(".local/share"));
    return QDir(base).filePath(APP_DIR_NAME);
#endif
}

SessionStore::SessionStore(const QString &stateDir)
    : m_stateDir(stateDir.isEmpty() ? defaultStateDir() : stateDir)
{}

QString SessionStore::path() const
{
    return QDir(m_stateDir).filePath(QLatin1StringView(FILENAME));
}

QString SessionStore::profileDir() const
{
    return QDir(m_stateDir).filePath(QStringLiteral("webprofile"));
}

void SessionStore::ensureDirs() const
{
    if (QDir(m_stateDir).exists())
        return;
    QDir().mkpath(m_stateDir);
    QFile::setPermissions(m_stateDir,
                          QFile::ReadOwner | QFile::WriteOwner | QFile::ExeOwner);
}

SessionState SessionStore::load() const
{
    QFile file(path());
    if (!file.open(QIODevice::ReadOnly))
        return {};

    QJsonParseError error{};
    const QJsonDocument doc = QJsonDocument::fromJson(file.readAll(), &error);
    if (error.error != QJsonParseError::NoError || !doc.isObject())
        return {};

    const QJsonObject raw = doc.object();
    const QJsonValue schema = raw.value(QStringLiteral("schema"));
    if (!schema.isDouble() || schema.toDouble() != SCHEMA_VERSION)
        return {};

    // Unknown keys from a future version are ignored; known ones are read.
    SessionState state;
    state.lastLogin = raw.value(QStringLiteral("last_login")).toDouble(0.0);
    state.lastSuccess = raw.value(QStringLiteral("last_success")).toDouble(0.0);
    state.lastExpiry = raw.value(QStringLiteral("last_expiry")).toDouble(0.0);
    state.lastTerm = raw.value(QStringLiteral("last_term")).toString();
    state.schema = SCHEMA_VERSION;
    return state;
}

void SessionStore::save(const SessionState &state) const
{
    ensureDirs();

    QJsonObject raw;
    raw.insert(QStringLiteral("last_login"), state.lastLogin);
    raw.insert(QStringLiteral("last_success"), state.lastSuccess);
    raw.insert(QStringLiteral("last_expiry"), state.lastExpiry);
    raw.insert(QStringLiteral("last_term"), state.lastTerm);
    raw.insert(QStringLiteral("schema"), state.schema);

    const QString target = path();
    const QString tmp = target + QStringLiteral(".tmp");
    {
        QFile file(tmp);
        if (!file.open(QIODevice::WriteOnly | QIODevice::Truncate)) {
            qCWarning(lcSession) << "could not write" << tmp << ":" << file.errorString();
            return;
        }
        file.setPermissions(QFile::ReadOwner | QFile::WriteOwner);
        file.write(QJsonDocument(raw).toJson(QJsonDocument::Indented));
    }
    QFile::setPermissions(tmp, QFile::ReadOwner | QFile::WriteOwner);

    // rename(2) replaces the target atomically; QFile::rename refuses to.
    if (std::rename(QFile::encodeName(tmp).constData(), QFile::encodeName(target).constData()) != 0)
        qCWarning(lcSession) << "could not replace" << target;
}

SessionState &SessionStore::markLogin(SessionState &state) const
{
    state.lastLogin = now();
    save(state);
    return state;
}

SessionState &SessionStore::markSuccess(SessionState &state) const
{
    state.lastSuccess = now();
    save(state);
    return state;
}

SessionState &SessionStore::markExpiry(SessionState &state) const
{
    state.lastExpiry = now();
    save(state);
    return state;
}

void SessionStore::clear() const
{
    QFile::remove(path());
}

} // namespace mycu
